lib/mini_racer/truffleruby.rb

```diff
@@ -0,0 +1,353 @@
+        convert_js_to_ruby(value)
+      end
+    end
+
+    def convert_js_to_ruby(value)
+      case value
+      when true, false, Integer, Float
+        value
+      else
+        if value.nil?
+          nil
+        elsif value.respond_to?(:call)
+          MiniRacer::JavaScriptFunction.new
+        elsif value.respond_to?(:to_str)
+          value.to_str
+        elsif value.respond_to?(:to_ary)
+          value.to_ary.map do |e|
+            if e.respond_to?(:call)
+              nil
+            else
+              convert_js_to_ruby(e)
+            end
+          end
+        elsif is_time(value)
+          js_date_to_time(value)
+        elsif is_symbol(value)
+          js_symbol_to_symbol(value)
+        else
+          object = value
+          h = {}
+          object.instance_variables.each do |member|
+            v = object[member]
+            unless v.respond_to?(:call)
+              h[member.to_s] = convert_js_to_ruby(v)
+            end
+          end
+          h
+        end
+      end
+    end
+    
+    def js_json_stringify(val)
+      @json_stringify_func ||= eval_in_context("(x) => { return JSON.stringify(x) };")
+      @json_stringify_func.call(val)
+    end
+
+    def is_object_or_array(val)
+      @is_object_or_array_func ||= eval_in_context("(x) => { return (x instanceof Object || x instanceof Array) && !(x instanceof Date) && !(x instanceof Function) };")
+      @is_object_or_array_func.call(val)
+    end
+
+    def is_time(value)
+      @is_time_func ||= eval_in_context("(x) => { return x instanceof Date };")
+      @is_time_func.call(value)
+    end
+
+    def js_date_to_time(value)
+      @js_date_to_time_func ||= eval_in_context("(x) => { return x.getTime(x) };")
+      millis = @js_date_to_time_func.call(value)
+      Time.at(Rational(millis, 1000))
+    end
+
+    def is_symbol(value)
+      @is_symbol_func ||= eval_in_context("(x) => { return typeof x === 'symbol' };")
+      @is_symbol_func.call(value)
+    end
+
+    def js_symbol_to_symbol(value)
+      @js_symbol_to_symbol_func ||= eval_in_context("(x) => { var r = x.description; return r === undefined ? 'undefined' : r };")
+      @js_symbol_to_symbol_func.call(value).to_sym
+    end
+
+    def js_new_date(value)
+      @js_new_date_func ||= eval_in_context("(x) => { return new Date(x) };")
+      @js_new_date_func.call(value)
+    end
+
+    def js_new_array(size)
+      @js_new_array_func ||= eval_in_context("(x) => { return new Array(x) };")
+      @js_new_array_func.call(size)
+    end
+
+    def js_set_value(object, key, value)
+      @js_set_value_func ||= eval_in_context("(o, k, v) => { o[k] = v };")
+      @js_set_value_func.call(object, key, value)
+    end
+
+    def convert_ruby_to_js(value)
+      case value
+      when nil, true, false, Integer, Float, String
+        value
+      when Array
+        ary = js_new_array(value.size)
+        value.each_with_index do |v, i|
+          ary[i] = convert_ruby_to_js(v)
+        end
```

Comment:
Using `js_new_array` works around the following issue:

I'm not sure why this array gets converted to `123` when calling the function with a ruby array:
```ruby
context = MiniRacer::Context.new
context.eval("function f(x, y) { console.log(x); return 'I need ' + x + ' ' + y }")
context.call('f', [1, 2, 3], 'bars')
# prints 123
```

Reply:
Sounds like a Graal.js issue worth fixing

Reply:
A graal.js issue was filed.